### PR TITLE
[reactive-element] Add standard decorators

### PR DESCRIPTION
**Pull request by justinfagnani**
Fixes https://github.com/lit/lit/issues/3200
Fixes https://github.com/lit/lit/issues/1476

TODO:

- [ ] Convert other packages like context. See https://github.com/lit/lit/pull/4009
- [x] Export decorators from `lit-element` and `lit`
- [x] Address https://github.com/lit/lit/issues/1476 (make initial values not reflect)
- [ ] Adjust jsdocs as necessary
- [x] Add `legacy-decorators.js` modules that re-export the existing decorators for forward-compat with Lit 4.0
- [x] Address the type errors that are currently `// @ts-expect-error: argh`'ed

Open questions:
- [x] Can we do better than adding a `newValue` parameter to `requestUpdate()`?
- [x] If not, can we remove it from the public API?
- [x] Should adding `@property()` on a setter wrap it and automatically call `requestUpdate()`? This would be a change from current decorators. If not, we should consider allowing decorating either the getter or setter for less of a breaking change.
- [x] Should we address other issues that require breaking changes, like https://github.com/lit/lit/issues/3714
- [x] Should we remove `@queryAsync()`? Does anyone use it? We don't have the same need for it internally.

### CODE REVIEW

**Comment by justinfagnani** [comment]
Got the core decorators building and test passing locally.

**Comment by justinfagnani** [comment]
We should address https://github.com/lit/lit/issues/1476 in this PR too

File: packages/reactive-element/src/std-decorators/property.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      return {
+        get(this: C) {
+          return (target as ClassAccessorDecoratorTarget<C, V>).get.call(this);
+        },
+        set(this: C, v: V) {
+          const oldValue = (
+            target as ClassAccessorDecoratorTarget<C, V>
+          ).get.call(this);
+          (target as ClassAccessorDecoratorTarget<C, V>).set.call(this, v);
+          this.requestUpdate(name, oldValue, options);
```
Comment (rictic) [inline packages/reactive-element/src/std-decorators/property.ts:85]: Need to check that the new value is different before calling requestUpdate, don't we? Use hasChanged if available?

**Review by rictic — CHANGES_REQUESTED** [review]
This throws a `Uncaught TypeError: Cannot read private member #foo_accessor_storage from an object whose class did not declare it`

```js
customElements.define(
  'foo-bar',
  class extends LitElement {
    @property() accessor foo = undefined;
  }
);
document.createElement('foo-bar');
```

File: packages/reactive-element/.gitignore [hunk 2]
```diff
@@ -1,9 +1,11 @@
 /decorators/
 /development/
 /node/
+/std-decorators/
 /test/
 /css-tag.*
 /decorators.*
+/std-decorators.*
 /polyfill-support.*
```
Comment (rictic) [inline packages/reactive-element/.gitignore:9]: I think we should also publish a /legacy-decorators.js file that just re-exports decorators

Comment (justinfagnani) [reply packages/reactive-element/.gitignore:11]: Thanks for doing this!

File: packages/reactive-element/src/reactive-element.ts [hunk 3]
```diff
@@ -737,6 +751,32 @@ export abstract class ReactiveElement
     return this.elementProperties.get(name) ?? defaultPropertyDeclaration;
   }
 
+  // This is a finalization step, but it needs to be separate from `finalize`
+  // because `finalize` can be called before a class has been fully initialized.
+  // This method should only be called once per class, as late as possible.
+  protected static collectMetadata() {
+    const metadata = this[Symbol.metadata];
+    if (metadata == null) {
+      return;
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:760]: Added dev mode warning

File: packages/reactive-element/src/std-decorators/property.ts (hunk 1 above)
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts:85]: `hasChanged` is the first thing that `requestUpdate` calls if a name is given.

File: packages/reactive-element/src/std-decorators/property.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        },
+        init(this: C, v: V): V {
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts:87]: I believe so, though in this current state it wouldn't matter since we only do anything if there is an initial value.

[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts:87]: We need a requestUpdate() call in here to add the property to the first update's changeProperties

File: packages/reactive-element/src/std-decorators/property.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+          // Save instance property through setter
+          if (this.hasOwnProperty(name)) {
+            // @ts-expect-error: argh
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts]: This code is removed. __initialize is already doing this

File: packages/reactive-element/src/std-decorators/property.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    properties.set(context.name, options);
+
+    if (kind === 'accessor') {
+      // Standard decorators cannot dynamically modify the class, so we can't
+      // replace a field with accessors. The user must use the new `accessor`
+      // keyword instead.
+      const {name} = context;
+      return {
+        get(this: C) {
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts:77]: yes!

File: packages/reactive-element/src/std-decorators/property.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+          return v;
+        },
+      };
+    } else if (kind === 'setter') {
+      // TODO: legacy decorators do not automatically call requestUpdate() like
+      // this, because it was difficult to wrape the user-written accessors.
+      // NOTE: Because we need to wrap the setter, and we can't modify the class
+      // directly in a standard decorator, we can only decorate setters, not
+      // getters. This is change from our legacy decorators.
+      // const {name} = context;
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts:104]: I commented this out because we need to decide if we want to wrap and automatically call requestUpdate like this.

I think it's good behavior, but it's a change from the legacy decorators. wdyt @kevinpschaaf ?

File: packages/reactive-element/src/reactive-element.ts [hunk 8]
```diff
@@ -737,6 +751,32 @@ export abstract class ReactiveElement
     return this.elementProperties.get(name) ?? defaultPropertyDeclaration;
   }
 
+  // This is a finalization step, but it needs to be separate from `finalize`
+  // because `finalize` can be called before a class has been fully initialized.
+  // This method should only be called once per class, as late as possible.
+  protected static collectMetadata() {
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:758]: made private

File: packages/reactive-element/src/reactive-element.ts [hunk 9]
```diff
@@ -588,6 +601,7 @@ export abstract class ReactiveElement
   static get observedAttributes() {
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:601]: We don't store the attributes anywhere right now, and to do so would add to retained memory, so I'd like to be conservative there.  __collectMetadata is idempotent, so this should be safe. I would like to simplify how we track element properties eventually maybe even soon, but it might be separate from this?

File: packages/reactive-element/src/reactive-element.ts [hunk 10]
```diff
@@ -498,7 +509,9 @@ export abstract class ReactiveElement
   private static __attributeToPropertyMap: AttributeMap;
 
   /**
-   * Marks class as having finished creating properties.
+   * Marks class as having been finalized, which includes creating properties
+   * from `static properties`, but does *not* include all properties created
+   * from decorators.
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:514]: done

File: packages/reactive-element/src/reactive-element.ts [hunk 11]
```diff
@@ -376,6 +376,17 @@ export type WarningKind =
 
 export type Initializer = (element: ReactiveElement) => void;
 
+// Ensure metadata is enabled...
+// TODO: remove
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+(Symbol as any).metadata ??= Symbol('metadata');
+
+// Maybe from a class's metadata object to property options
+export const propertyMetadata = new WeakMap<
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:385]: done

File: packages/labs/ssr-dom-shim/tsconfig.json [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     "target": "es2021",
-    "module": "ES2020",
+    "module": "NodeNext",
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/labs/ssr-dom-shim/tsconfig.json:4]: TS 5.2 forces us to align `module` and `moduleResolution`

File: packages/reactive-element/src/reactive-element.ts [hunk 13]
```diff
@@ -1145,15 +1185,22 @@ export abstract class ReactiveElement
...
     oldValue?: unknown,
-    options?: PropertyDeclaration
+    options?: PropertyDeclaration,
+    // This is needed because standard decorators can't call requestUpdate()
+    // after the field is defined and acceessing the field before it's
+    // defined is an error.
+    // TODO: Copy the reflection logic to the decorator, or hide this parameter
+    // from the public API.
+    newValue?: unknown,
+    reflect = true
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts]: not applicable anymore

File: packages/reactive-element/src/reactive-element.ts [hunk 14]
```diff
@@ -1163,7 +1210,11 @@ export abstract class ReactiveElement
         // Note, it's important that every change has a chance to add the
         // property to `_reflectingProperties`. This ensures setting
         // attribute + property reflects correctly.
-        if (options.reflect === true && this.__reflectingProperty !== name) {
+        if (
+          reflect &&
+          options.reflect === true &&
+          this.__reflectingProperty !== name
+        ) {
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:1231]: I added a `__propertyDefaults` field which records the default values. We write to that in the `init()` callback of the `@property` decorators.

A maybe big change in behavior is that I made attributeToProperty defer to that when an attribute is removed, and we don't even use the attribute converter. In well-behaved elements this should be fine - the converter would hopefully convert `null` to the default value anyway, but it is a change. I think this is good though.

I don't see the case where we need to remove it from `__reflectingProperties`. Can you expand on that?

File: packages/labs/cli/package.json [hunk 15]
```diff
@@ -87,7 +87,7 @@
     "command-line-commands": "^3.0.2",
     "command-line-usage": "^6.1.1",
     "tslib": "^2.0.3",
-    "typescript": "~5.1.6"
+    "typescript": "~4.9.0"
```
Comment (justinfagnani) [inline packages/labs/cli/package.json:90]: I changed this when trying to resolve unmet peer dependencies a while back. I'll try to change it forward, though I think I'd rather stick with an older version than do a force install, since violated constraints break some npm commands.

File: packages/reactive-element/src/reactive-element.ts [hunk 16]
```diff
@@ -737,6 +763,31 @@ export abstract class ReactiveElement
     return this.elementProperties.get(name) ?? defaultPropertyDeclaration;
   }
 
+  // This is a finalization step, but it needs to be separate from `finalize`
```
Comment (sorvell) [inline packages/reactive-element/src/reactive-element.ts:766]: I don't see why this must be separate from `finalize` and doing so adds complexity. 

1. doesn't need to be separate: although `finalize` can be called more than once, it immediately returns if it's already done its work. 

2. complexity: calling finalize also ensures that superclasses have also been finalized, and separating this metadata from that means that a superclass may not have metadata. I'm not sure if that matters or not.

File: packages/reactive-element/src/std-decorators/property.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/reactive-element/src/std-decorators/property.ts:113]: I'm concerned about the perf implications and complexity of using `queueMicrotask` here, especially since there's already an initial call to `requestUpdate`. 

Instead, it seems like populating `changedProperties` with the defaults on first rendered can just be handled in the first call to `update`.

File: packages/reactive-element/src/reactive-element.ts [hunk 17]
```diff
@@ -1161,7 +1242,11 @@ export abstract class ReactiveElement
         // Note, it's important that every change has a chance to add the
         // property to `_reflectingProperties`. This ensures setting
         // attribute + property reflects correctly.
-        if (options.reflect === true && this.__reflectingProperty !== name) {
+        if (
+          !initial &&
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/reactive-element/src/reactive-element.ts:1246]: Note also that we are not matching platform behavior here precisely wrt reflecting properties. For example, any set to `input.type` will reflect, including not changing the property from its default. 

The code here does not match that since `requestUpdate` immediately returns if there is no value change.

File: packages/reactive-element/src/reactive-element.ts [hunk 18]
```diff
@@ -961,14 +1020,23 @@ export abstract class ReactiveElement
...
-    for (const p of elementProperties.keys()) {
+    for (const p of elementProperties.keys() as IterableIterator<keyof this>) {
       if (this.hasOwnProperty(p)) {
-        this.__instanceProperties!.set(p, this[p as keyof this]);
-        delete this[p as keyof this];
+        instanceProperties.push([p, this[p]]);
+        delete this[p];
       }
     }
+    if (instanceProperties.length) {
```
Comment (sorvell) [inline packages/reactive-element/src/reactive-element.ts:1033]: Is this change necessary?

I'm concerned about the potential perf implications (but hopefully this path shouldn't be hit a lot) and additional complexity (e.g. the comment in `_initialize` about ordering wrt `requestUpdate`).

Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:1033]: This isn't completely necessary, but it eliminates the __instanceProperties field and it's gated by there being instance properties to save. Once standard decorators are the only way to define fields we can do this w/o a microtask in an addInitializer() callback.

Comment (sorvell) [reply packages/reactive-element/src/reactive-element.ts:1033]: Per chat, not strictly necessary but preferred since it saves memory and centralizes the code, and this is only called in hopefully a corner case.

File: packages/reactive-element/src/std-decorators/property.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts:114]: Directly populating changeProperties now

File: packages/reactive-element/src/reactive-element.ts [hunk 19]
```diff
@@ -737,6 +763,31 @@ export abstract class ReactiveElement
...
+    }
+    const properties = litPropertyMetadata.get(metadata);
+    if (properties === undefined) {
+      return;
+    }
+    for (const [p, options] of properties) {
+      this.elementProperties.set(p, options);
+      // If this class doesn't have its own set, create one and initialize
+      // with the values in the set from the nearest ancestor class, if any.
+      if (!this.hasOwnProperty('__reactivePropertyKeys')) {
```
Comment (sorvell) [inline packages/reactive-element/src/reactive-element.ts:782]: I believe this is exclusively for a DEV_MODE warning so should be gated for that setting.

File: packages/reactive-element/src/reactive-element.ts (hunk 16 above)
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:766]: Talked offline: this needs to happen later than finalize(), which can be called early in addInitializer() and createProperty(). finalize() should probably be called prepare() in those cases.

File: packages/reactive-element/src/reactive-element.ts (hunk 19 above)
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts:782]: done

File: packages/reactive-element/src/reactive-element.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts]: removed initial

[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/reactive-element.ts]: removed `initial`

File: packages/reactive-element/src/std-decorators/property.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rictic) [reply packages/reactive-element/src/std-decorators/property.ts:132]: Decided that we don't want to call requestUpdate, it's the setter's responsibility

File: packages/reactive-element/src/std-decorators/query-assigned-elements.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+  ): ClassAccessorDecoratorResult<C, V> => {
+    const {slot, selector} = options ?? {};
+    return {
+      get(this: C): V {
+        const slotSelector = `slot${slot ? `[name=${slot}]` : ':not([name])'}`;
+        const slotEl =
+          this.renderRoot?.querySelector<HTMLSlotElement>(slotSelector);
+        const elements = slotEl?.assignedElements(options) ?? [];
+        if (selector) {
+          // @ts-expect-error: argh!
```
Comment (rictic) [inline packages/reactive-element/src/std-decorators/query-assigned-elements.ts:74]: Can't we cast here instead?

File: packages/reactive-element/src/std-decorators/query-assigned-nodes.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+ *     `;
+ *   }
+ * }
+ * ```
+ *
+ * Note the type of this property should be annotated as `Array<Node>`. Use the
+ * queryAssignedElements decorator to list only elements, and optionally filter
+ * the element list using a CSS selector.
+ *
+ * @category Decorator
```
Comment (rictic) [inline packages/reactive-element/src/std-decorators/query-assigned-nodes.ts:51]: todo: infer more precise return type when the query is a tagname

File: packages/reactive-element/src/std-decorators/query.ts [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ *   render() {
+ *     return html`
+ *       <div id="first"></div>
+ *       <div id="second"></div>
+ *     `;
+ *   }
+ * }
+ * ```
+ * @category Decorator
```
Comment (rictic) [inline packages/reactive-element/src/std-decorators/query.ts:39]: todo: infer more precise return type when the query is a tagname

File: packages/reactive-element/tsconfig.json [hunk 23]
```diff
@@ -3,7 +3,7 @@
     "composite": true,
     "target": "es2021",
     "module": "NodeNext",
-    "lib": ["es2021", "DOM", "DOM.Iterable"],
+    "lib": ["es2021", "DOM", "DOM.Iterable", "esnext.decorators"],
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/tsconfig.json:6]: it's needed for the `.metadata` field on the decorator context.

File: packages/reactive-element/src/test/std-decorators/property_test.ts [hunk 24]
```diff
@@ -0,0 +1,546 @@
...
+    await el.updateComplete;
+    assert.equal(el.all, 5);
+    assert.equal(el.updateCount, 5);
+    el.all = 15;
+    await el.updateComplete;
+    assert.equal(el.all, 15);
+    assert.equal(el.updateCount, 6);
+  });
+
+  // This test isn't valid. You can not override a field that's initialized
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/test/std-decorators/property_test.ts:290]: This is a limitation is JS now. You can't do:

```ts
class A {
  constructor() {
    this.foo = 1;
  }
}
class B extends A {
  accessor foo;
}
```

because A's `this.foo = 1` runs before B's foo accessor has it's private storage defined. That's what this test is trying to do, so at least that bit is not even testable. The main idea that a subclass can't interfere with a superclass it good though... I'm rewriting to avoid the timing problem.

File: packages/reactive-element/src/test/std-decorators/property_test.ts [hunk 25]
```diff
@@ -0,0 +1,546 @@
...
+    assert.isTrue(changedProperties!.has('foo'));
+
+    // Check that the element is still reactive
+    changedProperties = undefined;
+    el.foo = 'bye';
+    await el.updateComplete;
+    assert.isTrue(changedProperties!.has('foo'));
+  });
+
+  // We can not support custom property descriptors anymore. This use case
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/test/std-decorators/property_test.ts:415]: done

File: packages/reactive-element/src/std-decorators/state.ts [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Properties declared this way must not be used from HTML or HTML templating
+ * systems, they're solely for properties internal to the element. These
+ * properties may be renamed by optimization tools like closure compiler.
+ * @category Decorator
+ */
+export function state(
+  options?: InternalPropertyDeclaration
+): PropertyDecorator {
+  return property({
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/state.ts:38]: I think the cost is low. The original options object will be very short-lived, in the young generation of generational GCs and cheap to collect.

File: packages/reactive-element/src/std-decorators/property.ts [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+  // setter decorator signature
+  <C extends ReactiveElement, V>(
+    target: (value: V) => void,
+    context: ClassSetterDecoratorContext<C, V>
+  ): (this: C, value: V) => void;
+};
+
+// TODO (justinfagnani): export from ReactiveElement?
+// It actually makes sense to have this default defined with the decorator, so
+// that different decirators could have different defaults.
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/property.ts:60]: done

File: packages/reactive-element/src/std-decorators/query-assigned-elements.ts [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+ *       <slot></slot>
+ *     `;
+ *   }
+ * }
+ * ```
+ *
+ * Note, the type of this property should be annotated as `Array<HTMLElement>`.
+ *
+ * @category Decorator
+ */
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/query-assigned-elements.ts:59]: Done. I played around with this and it's quite tricky

File: packages/reactive-element/src/std-decorators/query-assigned-elements.ts (hunk 20 above)
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/query-assigned-elements.ts:74]: Yes, I was hoping to find out why `as unknown as` was required, since V should extend Array<Element>, and I thought that should work with a simple cast.

File: packages/reactive-element/src/std-decorators/query-assigned-nodes.ts (hunk 21 above)
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/query-assigned-nodes.ts:51]: done

File: packages/reactive-element/src/std-decorators/query.ts (hunk 22 above)
Comment (justinfagnani) [reply packages/reactive-element/src/std-decorators/query.ts:39]: done

**Comment by justinfagnani** [comment]
@rictic I forgot to push my changes last night - just pushed now.

**Review by rictic — APPROVED** [review]
